llvm/lib/Transforms/Utils/SYCLModuleSplit.cpp

```diff
@@ -0,0 +1,935 @@
+// global variables) in a module. It is used during device code split to
+// understand which global variables and functions (other than entry points)
+// should be included into a split module.
+//
+// Nodes of the graph represent LLVM's GlobalObjects, edges "A" -> "B" represent
+// the fact that if "A" is included into a module, then "B" should be included
+// as well.
+//
+// Examples of dependencies which are represented in this graph:
+// - Function FA calls function FB
+// - Function FA uses global variable GA
+// - Global variable GA references (initialized with) function FB
+// - Function FA stores address of a function FB somewhere
+//
+// The following cases are treated as dependencies between global objects:
+// 1. Global object A is used within by a global object B in any way (store,
+//    bitcast, phi node, call, etc.): "A" -> "B" edge will be added to the
+//    graph;
+// 2. function A performs an indirect call of a function with signature S and
+//    there is a function B with signature S marked with "referenced-indirectly"
+//    attribute. "A" -> "B" edge will be added to the graph;
+class DependencyGraph {
+public:
+  using GlobalSet = SmallPtrSet<const GlobalValue *, 16>;
+
+  DependencyGraph(const Module &M) {
+    // Group functions by their signature to handle case (2) described above
+    DenseMap<const FunctionType *, DependencyGraph::GlobalSet>
+        FuncTypeToFuncsMap;
+    for (const auto &F : M.functions()) {
+      // Kernels can't be called (either directly or indirectly) in SYCL
+      if (isKernel(F))
+        continue;
+
+      // Only functions which are marked with "referenced-indireclty" attribute
+      // are considered to be indirect callee candidates.
+      if (!F.hasFnAttribute("referenced-indirectly"))
+        continue;
+
+      FuncTypeToFuncsMap[F.getFunctionType()].insert(&F);
+    }
+
+    for (const auto &F : M.functions()) {
+      // case (1), see comment above the class definition
+      for (const Value *U : F.users())
+        addUserToGraphRecursively(cast<const User>(U), &F);
+
+      // case (2), see comment above the class definition
+      for (const auto &I : instructions(F)) {
+        const auto *CI = dyn_cast<CallInst>(&I);
+        if (!CI || !CI->isIndirectCall()) // Direct calls were handled above
+          continue;
+
+        // TODO: consider limiting set of potential callees to functions marked
+        // with special attribute (like [[intel::device_indirectly_callable]])
+        const FunctionType *Signature = CI->getFunctionType();
+        // Note: strictly speaking, virtual functions are allowed to use
+        // co-variant return types, i.e. we can actually miss a potential callee
+        // here, because it has different signature (different return type).
+        // However, this is not a problem for two reasons:
+        // - opaque pointers will be enabled at some point and will make
```

Comment:
That's my bad for not updating the comments, but opaque pointers have been enabled at intel/llvm already and they had been enabled even earlier in the upstream. This should say something like "opaque pointers will make those signatures the same"